File: stack/evaluate_reverse_polish_notation/js/evaluate_reverse_polish_notation.js
const tokens = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"];

const evalRPN = (tokens) => {
  let stack = [];

  for(let s of tokens) {
    let n = parseInt(s);
    if(!Number.isNaN(n)) {
      stack.push(n);
      continue;
    }
    let b = stack.pop();
    let a = stack.pop();

    switch(s) {
      case "+":
        stack.push(a + b);
        break;
      case "-":
        stack.push(a - b);
        break;
      case "*":
        stack.push(a * b);
        break;
      case "/":
        stack.push(Math.trunc(a / b));
        break;
      default:
        throw new Error('Must be operator or number');
    }
  }
  return stack[0];
};

console.log(evalRPN(tokens));

File: stack/evaluate_reverse_polish_notation/cpp/evaluate_reverse_polish_notation.cpp
#include<iostream>
#include<stdexcept>
#include<vector>
#include<string>
#include<stack>

using std::cout;
using std::vector;
using std::string;

// Time O(n)
// Space O(n)

class Solution {
public:
    static int evalRPN(const vector<string>& tokens) {
        std::stack<int> stack;

        for(const auto& s : tokens) {
            // I originally implimented with try catch(...)
            // the try catch was 8x slower, checking with isdigit
            // made it much faster 90ms to 11ms
            if(s.size() > 1 || std::isdigit(s[0])) {
                stack.push(std::stoi(s));
                continue;
            }
            char c = s[0];

            int b = stack.top();
            stack.pop();
            int a = stack.top();
            stack.pop();

            switch(c) {
                case '+':
                    stack.push(a + b);
                    break;
                case '-':
                    stack.push(a - b);
                    break;
                case '*':
                    stack.push(a * b);
                    break;
                case '/':
                    stack.push(a / b);
                    break;
                default:
                    throw std::invalid_argument("Must be operator or int");
            }
        }
        return stack.top();
    }
};

int main() {
    vector<string> tokens = { "10","6","9","3","+","-11","*","/","*","17","+","5","+" };

    int answer = Solution::evalRPN(tokens);

    cout << answer << '\n';
}
